src/main/java/net/minecraftforge/fml/relauncher/DependencyExtractor.java

```diff
@@ -0,0 +1,344 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.fml.relauncher;
+
+import com.google.common.base.Charsets;
+import com.google.common.collect.ObjectArrays;
+import com.google.common.io.ByteStreams;
+import com.google.common.io.Files;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonSyntaxException;
+import net.minecraftforge.fml.common.FMLLog;
+import net.minecraftforge.fml.relauncher.ModListHelper.JsonModList;
+import org.apache.commons.io.IOUtils;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.jar.Attributes;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.jar.JarInputStream;
+import java.util.jar.Manifest;
+import java.util.stream.Collectors;
+
+public class DependencyExtractor
+{
+    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
+    private static final Attributes.Name MOD_CONTAINS_DEPS = new Attributes.Name("ContainedDeps");
+    private static final Attributes.Name MAVEN_ARTIFACT = new Attributes.Name("MavenArtifact");
+    private static final List<String> skipContainedDeps = Arrays.asList(System.getProperty("fml.skipContainedDeps", "").split(","));
+    private static int extractedDeps = 0;
+
+    public static void inspect(File baseModsDir, File versionedModsDir)
+    {
+        FMLLog.log.debug("Inspecting mod directory for dependency extraction candidates");
+        File[] modFiles = baseModsDir.listFiles((dir, name) -> name.endsWith(".jar"));
+        if (versionedModsDir.isDirectory())
+        {
+            File[] versionedMods = versionedModsDir.listFiles((dir, name) -> name.endsWith(".jar"));
+            modFiles = ObjectArrays.concat(modFiles, versionedMods, File.class);
+        }
+        File modListFile = new File(baseModsDir, "mod_list.json");
+        JsonModList modList = prepareModList(modListFile);
+        File repository = new File(modList.repositoryRoot);
+        // It's a little hacky, but it should yield the best performance
+        Map<Artifact, Artifact> artifacts = modList.modRef.stream().map(Artifact::new).collect(Collectors.toMap(a -> a, a -> a));
```

Comment:
The "hacky" part is the map. I'm really using it like a set with easy retrieval of existing elements. The stream performance shouldn't matter too much here, since this only happens once at startup.